File: implementations/java/src/main/java/com/webweavex/Graph.java
package com.webweavex;

import java.util.*;

public class Graph {
    
    public Map<String, Object> build(List<Map<String, String>> entities) {
        Map<String, Object> result = new LinkedHashMap<>();
        
        List<Map<String, Object>> nodes = new ArrayList<>();
        List<Map<String, Object>> edges = new ArrayList<>();
        
        if (entities == null || entities.isEmpty()) {
            result.put("nodes", nodes);
            result.put("edges", edges);
            return result;
        }
        
        List<Map<String, String>> uniqueEntities = new ArrayList<>();
        Set<String> seen = new HashSet<>();
        
        for (Map<String, String> e : entities) {
            String key = e.get("type") + ":" + e.get("value");
            if (!seen.contains(key)) {
                seen.add(key);
                uniqueEntities.add(e);
            }
        }
        
        Collections.sort(uniqueEntities, (a, b) -> {
            int typeComp = a.get("type").compareTo(b.get("type"));
            if (typeComp != 0) return typeComp;
            return a.get("value").compareTo(b.get("value"));
        });
        
        for (Map<String, String> e : uniqueEntities) {
            Map<String, Object> node = new LinkedHashMap<>();
            String id = e.get("type") + ":" + e.get("value");
            node.put("id", id);
            node.put("type", e.get("type"));
            node.put("value", e.get("value"));
            nodes.add(node);
        }
        
        for (int i = 0; i < uniqueEntities.size(); i++) {
            for (int j = i + 1; j < uniqueEntities.size(); j++) {
                Map<String, Object> edge = new LinkedHashMap<>();
                edge.put("source", uniqueEntities.get(i).get("type") + ":" + uniqueEntities.get(i).get("value"));
                edge.put("target", uniqueEntities.get(j).get("type") + ":" + uniqueEntities.get(j).get("value"));
                edge.put("weight", 1);
                edges.add(edge);
            }
        }
        
        nodes.sort((a, b) -> ((String) a.get("id")).compareTo((String) b.get("id")));
        edges.sort((a, b) -> {
            int sourceComp = ((String) a.get("source")).compareTo((String) b.get("source"));
            if (sourceComp != 0) return sourceComp;
            return ((String) a.get("target")).compareTo((String) b.get("target"));
        });
        
        result.put("nodes", nodes);
        result.put("edges", edges);
        return result;
    }
}

File: implementations/kotlin/src/main/kotlin/com/webweavex/Graph.kt
package com.webweavex

data class GraphNode(
    val id: String,
    val type: String,
    val value: String
)

data class GraphEdge(
    val source: String,
    val target: String,
    val weight: Int = 1
) {
    fun toMap(): Map<String, Any> = mapOf("source" to source, "target" to target, "weight" to weight)
}

data class GraphData(
    val nodes: List<GraphNode>,
    val edges: List<GraphEdge>
)

class Graph {
    fun build(entities: List<Entity>): GraphData {
        if (entities.isEmpty()) {
            return GraphData(emptyList(), emptyList())
        }

        val uniqueEntities = entities.distinct()

        val sortedEntities = uniqueEntities.sortedWith(
            compareBy<Entity> { it.type }.thenBy { it.value }
        )

        val nodes = sortedEntities.map { entity ->
            val id = "${entity.type}:${entity.value}"
            GraphNode(id, entity.type, entity.value)
        }.sortedBy { it.id }

        val edges = mutableListOf<GraphEdge>()
        for (i in sortedEntities.indices) {
            for (j in i + 1 until sortedEntities.size) {
                val source = "${sortedEntities[i].type}:${sortedEntities[i].value}"
                val target = "${sortedEntities[j].type}:${sortedEntities[j].value}"
                edges.add(GraphEdge(source, target, 1))
            }
        }

        val sortedEdges = edges.sortedWith(
            compareBy<GraphEdge> { it.source }.thenBy { it.target }
        )

        return GraphData(nodes, sortedEdges)
    }
}
